feat(ivy): SVG handling added

benlesh opened the pull request:
Initial work for adding SVG handling to Ivy.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

Adds new features to ivy render regarding namespaces and SVG

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?

SVG and namespaces are now handled by Ivy renderer

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

Given that the array of attributes being passed around is now `Array<string | NS>`, it might be worth looking into an abstraction to loop over that array, since we're doing it in many places. For now I'm just doing that loop in-line.

mary-poppins commented:
You can preview c806390 at https://pr23899-c806390.ngbuilds.io/.

mary-poppins commented:
You can preview e01e1cd at https://pr23899-e01e1cd.ngbuilds.io/.

mary-poppins commented:
You can preview d8ba71e at https://pr23899-d8ba71e.ngbuilds.io/.

JoostK commented on packages/core/src/render3/instructions.ts:
```diff
@@ -956,6 +996,15 @@ export function elementEnd() {
 }
 
+/** Marks the beginning and end of an element in one call. */
+export function element(
```
Is the compiler expected to generate this instruction in favor of `elementStart`/`elementEnd`?

pkozlowski-opensource replied:
This new instruction is here to be introduced / stay but should probably be done in a separate PR with dedicated unit tests, compiler changes etc.

benlesh replied:
@mhevery  done.

JoostK commented on packages/core/src/render3/instructions.ts:
```diff
@@ -812,15 +842,25 @@ export function createTView(
+      const attrName = attrs[i++] as string;
+      const attrVal = attrs[i++] as string;
+      isProc ? (renderer as ProceduralRenderer3).setAttributeNS(native, attrNS, attrName, attrVal) :
+               native.setAttribute(attrName, attrVal);
```
Looks like this needs to be a call into `setAttributeNS`. According to MDN the attribute name should be FQ'd including namespace alias, which we currently do not account for it seems.

JoostK commented on packages/core/src/render3/instructions.ts:
```diff
@@ -956,6 +996,15 @@ export function elementEnd() {
+/** Marks the beginning and end of an element in one call. */
+export function element(
+    index: number, name: string, attrs?: string[] | null | undefined,
+    localRefs?: string[] | null | undefined) {
```
Can we specify `RElement` to be the return type, if this instruction is here to stay.

mary-poppins commented:
You can preview 01dccc9 at https://pr23899-01dccc9.ngbuilds.io/.

mary-poppins commented:
You can preview 990ec57 at https://pr23899-990ec57.ngbuilds.io/.

mary-poppins commented:
You can preview 931b529 at https://pr23899-931b529.ngbuilds.io/.

mary-poppins commented:
You can preview d7b76b6 at https://pr23899-d7b76b6.ngbuilds.io/.

mary-poppins commented:
You can preview b4ef8b7 at https://pr23899-b4ef8b7.ngbuilds.io/.

mary-poppins commented:
You can preview cfa503d at https://pr23899-cfa503d.ngbuilds.io/.

mary-poppins commented:
You can preview 6a3ae23 at https://pr23899-6a3ae23.ngbuilds.io/.

mary-poppins commented:
You can preview a4faf1f at https://pr23899-a4faf1f.ngbuilds.io/.

mary-poppins commented:
You can preview 3a22b02 at https://pr23899-3a22b02.ngbuilds.io/.

pkozlowski-opensource commented on packages/core/test/bundling/todo/todo_e2e_spec.ts:
```diff
@@ -27,6 +27,7 @@ describe('functional test for todo', () => {
            expect(document.body.textContent).toContain('todos');
            expect(document.body.textContent).toContain('Demonstrate Components');
            expect(document.body.textContent).toContain('4 items left');
+           debugger;
```
`debugger;` left

pkozlowski-opensource commented on packages/core/test/render3/instructions_spec.ts:
```diff
@@ -355,6 +355,43 @@ describe('instructions', () => {
+          'height',
+          '300',
+          // test:title="abc"
+          NS.FULL,
```
Could you please add a test for a case where we've got multiple namespaced attributes? From this one case it is not clear if the `NS.FULL`  marker should be present only once or should it be "prefixing" each and every namespaced attribute.

benlesh replied:
done

pkozlowski-opensource commented on packages/core/src/render3/instructions.ts:
```diff
@@ -561,6 +561,39 @@ function getRenderFlags(view: LView): RenderFlags {
+  XLink: 'http://www.w3.org/1999/xlink',
+};
+
+export enum NS {
```
This should be probably re-exported from `index.ts` as this marker will show up in the generated code.

pkozlowski-opensource commented on packages/core/test/render3/outputs_spec.ts:
```diff
@@ -87,10 +87,10 @@ describe('outputs', () => {
     renderToHtml(Template, ctx, deps);
 
-    buttonToggle !.change.next();
+    buttonToggle !.change.emit();
```
If I'm not mistaken this is not related to the current PR? Maybe it would be better to have those changes in a separate refactor PR?

mary-poppins commented:
You can preview bf0b0ac at https://pr23899-bf0b0ac.ngbuilds.io/.

mary-poppins commented:
You can preview 6f40a90 at https://pr23899-6f40a90.ngbuilds.io/.

mary-poppins commented:
You can preview 7763b12 at https://pr23899-7763b12.ngbuilds.io/.

mary-poppins commented:
You can preview 321e23f at https://pr23899-321e23f.ngbuilds.io/.

pkozlowski-opensource commented on packages/core/test/render3/instructions_spec.ts:
```diff
@@ -355,6 +355,44 @@ describe('instructions', () => {
+
+      expect([standardHTML, ie11HTML]).toContain(t.html);
+    });
+  });
```
I think that we should have more tests as the `attrs` array is used in several places:
* setting an attribute in the DOM
* setting an input of a directive from an attribute
* matching directive selector
* matching nodes for content projection
* injecting attributes with `@Attribute`

IMO we should have at least one test for each of the mentioned cases.

mhevery replied:
👍

mary-poppins commented:
You can preview 8ea0182 at https://pr23899-8ea0182.ngbuilds.io/.

mary-poppins commented:
You can preview 4b70b22 at https://pr23899-4b70b22.ngbuilds.io/.

mary-poppins commented:
You can preview 68b303c at https://pr23899-68b303c.ngbuilds.io/.

mary-poppins commented:
You can preview 07956bd at https://pr23899-07956bd.ngbuilds.io/.

mary-poppins commented:
You can preview 262a424 at https://pr23899-262a424.ngbuilds.io/.

mary-poppins commented:
You can preview a32e951 at https://pr23899-a32e951.ngbuilds.io/.

mary-poppins commented:
You can preview 49f428a at https://pr23899-49f428a.ngbuilds.io/.

mary-poppins commented:
You can preview 5e2b516 at https://pr23899-5e2b516.ngbuilds.io/.

mary-poppins commented:
You can preview 777f906 at https://pr23899-777f906.ngbuilds.io/.

benlesh commented on packages/core/src/render3/instructions.ts:
```diff
@@ -563,6 +563,39 @@ function getRenderFlags(view: LView): RenderFlags {
+}
+
+export function setHtmlNS() {
+  _currentNS = KnownNS.XHTML;
```
Done

benlesh replied:
Why null? Then we have to check it every time it's used. I just set it to the appropriate string inlined.

benlesh commented on packages/core/src/render3/instructions.ts:
```diff
@@ -563,6 +563,39 @@ function getRenderFlags(view: LView): RenderFlags {
+}
+
+export function setSvgNS() {
+  _currentNS = KnownNS.SVG;
```
Done

benlesh replied:
inlined.

benlesh commented on packages/core/src/render3/interfaces/node.ts:
```diff
@@ -6,6 +6,8 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
+import {NS} from '../instructions';
```
done

benlesh replied:
done

mary-poppins commented:
You can preview 40debe9 at https://pr23899-40debe9.ngbuilds.io/.

benlesh commented on packages/core/src/render3/interfaces/renderer.ts:
```diff
@@ -16,7 +16,7 @@
 
 import {ViewEncapsulation} from '../../metadata/view';
-import {RendererStyleFlags2, RendererType2} from '../../render/api';
+import {Renderer2, RendererStyleFlags2, RendererType2} from '../../render/api';
```
done.

benlesh replied:
done.

benlesh commented on packages/core/src/render3/instructions.ts:
```diff
@@ -563,6 +563,39 @@ function getRenderFlags(view: LView): RenderFlags {
+//// Namespace
+//////////////////////////
+export const KnownNS = {
+  XHTML: 'http://www.w3.org/1999/xhtml',
```
I removed it.

mary-poppins commented:
You can preview c01a7ba at https://pr23899-c01a7ba.ngbuilds.io/.

mary-poppins commented:
You can preview 1009c80 at https://pr23899-1009c80.ngbuilds.io/.

mary-poppins commented:
You can preview d271a99 at https://pr23899-d271a99.ngbuilds.io/.

mhevery commented on packages/compiler/src/render3/r3_identifiers.ts:
```diff
@@ -21,6 +21,8 @@ export class Identifiers {
 
```
We talked about renaming `createEleement` to `elementStart` to be consistent. It should be a minor change.

benlesh replied:
done

mhevery commented on packages/core/src/render3/instructions.ts:
```diff
@@ -563,6 +563,28 @@ function getRenderFlags(view: LView): RenderFlags {
+}
+
+export function setHtmlNS() {
+  _currentNS = 'http://www.w3.org/1999/xhtml';
```
I left a comment previously about making the `HTML` strategy `null`. Did you decide against it or forgot?
See perf https://jsperf.com/createelement-vs-createelementns of 24% penalty

benlesh replied:
done

benlesh replied:
done

mhevery commented on packages/core/src/render3/instructions.ts:
```diff
@@ -580,13 +613,18 @@ function getRenderFlags(view: LView): RenderFlags {
-  const native: RElement = renderer.createElement(name);
+
+  const native: RElement = isProceduralRenderer(renderer) ?
+      renderer.createElement(name) :
```
please address the comment.

benlesh committed:
feat(ivy): add namespace instructions for SVG and others

vicb commented on packages/compiler/src/render3/view/template.ts:
```diff
@@ -313,32 +313,42 @@ export class TemplateDefinitionBuilder implements t.Visitor<void>, LocalResolver
-    this.instruction(
-        this._creationCode, element.sourceSpan, R3.createElement, ...trimTrailingNulls(parameters));
+
+    let isSelfClosingElement = element.outputs.length === 0 && element.children.length === 0;
```
nits:
- let -> const
- "self closing" elements are a different thing. Change the name ?

benlesh replied:
change to `const isEmptyElement`

vicb commented on packages/compiler/test/render3/r3_view_compiler_template_spec.ts:
```diff
@@ -109,4 +109,43 @@ describe('compiler compliance: template', () => {
+          }
+        }`;
+
+      debugger;
```
oops

benlesh replied:
fixed.

vicb commented on packages/compiler/test/render3/r3_view_compiler_template_spec.ts:
```diff
@@ -109,4 +109,43 @@ describe('compiler compliance: template', () => {
+      };
+
+      const template = `
+        // ...
```
remove this. Expectations are not anchored.

vicb commented on packages/core/src/render3/instructions.ts:
```diff
@@ -819,17 +848,29 @@ export function createTView(
+  while (i < attrs.length) {
+    const first = attrs[i++];
+    ngDevMode && ngDevMode.rendererSetAttribute++;
+    if (first === 0) {  // NS.FULL
```
Why not using `Namespace.Full` instead ?

benlesh replied:
Updated to use `AttrubuteMarker.NAMESPACE_URI`

vicb commented on packages/core/test/render3/instructions_spec.ts:
```diff
@@ -355,6 +391,101 @@ describe('instructions', () => {
+      const ieHTML =
+          '<div id="container" style="background: rgb(222, 173, 17);" title="abc" whatever="wee" shazbot="wocka wocka"></div>';
+
+      expect([standardHTML, ieHTML]).toContain(t.html);
```
IIUC this test does not check the namespaces of attribute ?

benlesh replied:
Added some checks.

vicb reviewed (changes requested):
- See inline comments,
- There should be some compiler canonical tests (core/test/render3/compiler_canonical)
- Seems like the compiler is not fully updated to support namespaces ? ie `SetNS` and the like are never generated. Please create a github issue with what should be done.
- Update `core/src/render3/STATUS.md` with what is done and what is left to be done (issue created as indicated in the previous bullet)

benlesh commented on packages/core/src/render3/interfaces/node.ts:
```diff
@@ -14,6 +13,16 @@ import {RElement, RNode, RText} from './renderer';
+/**
+ * Namespace attribute flags.
+ */
+export const enum NS {
```
Updated to use `AttrubuteMarker.NAMESPACE_URI`, this enum has been removed.

benlesh commented on packages/core/src/render3/node_selector_matcher.ts:
```diff
@@ -98,10 +99,18 @@ function isPositive(mode: SelectorFlags): boolean {
   for (let i = 0; i < attrs.length; i += 2) {
-    if (attrs[i] === name) return i;
+    const first = attrs[i];
+    if (first === 0) {  // NS.FULL
```
Done.

mary-poppins commented:
You can preview 8e1119f at https://pr23899-8e1119f.ngbuilds.io/.

mary-poppins commented:
You can preview 8cb8ec5 at https://pr23899-8cb8ec5.ngbuilds.io/.

mary-poppins commented:
You can preview 134e28b at https://pr23899-134e28b.ngbuilds.io/.

vicb commented on packages/compiler/test/render3/r3_view_compiler_template_spec.ts:
```diff
@@ -109,4 +109,40 @@ describe('compiler compliance: template', () => {
     expectEmit(result.source, template, 'Incorrect template');
   });
 
+  // tslint:disable-next-line:ban
```
curious what was the issue with that next line ?

vicb commented on packages/core/src/render3/interfaces/node.ts:
```diff
@@ -163,7 +160,11 @@ export interface LProjectionNode extends LNode {
+   * Use the next value as the full namespaces URI, the values after that
+   * are then the name and the value, respectively.
+   */
+  NAMESPACE_URI = 0,  // namespace. Has to be repeated.
```
Should be `NamespaceUri`

benlesh replied:
That doesn't seem right, `SELECT_ONLY` is right below it. Should that change too?

vicb replied:
Yep, it should change too

vicb commented on packages/core/src/render3/STATUS.md:
```diff
@@ -150,7 +150,7 @@ The goal is for the `@Component` (and friends) to be the compiler of template. S
 | `{{ { a: 'literal', b: exp } }}`            |  ✅     |  ✅      |  ✅      |
 | `{{ exp \| pipe: arg }}`                    |  ✅     |  ✅      |  ✅      |
-| `<svg:g svg:p>`                             |  ❌     |  ❌      |  ❌      |
+| `<svg:g svg:p>`                             |  ✅     |  ✅      |  ❌      |
```
I think spec (2nd column) here is for compiler spec ? @mhevery ?

Anyway this compiler spec does not exist yet, right ? (doc or gh issue)

mhevery replied:
I thought this PR would also do the changes to the compiler to generate the `namespaceSVG()` calls. @benlesh could you add that so we can complete the row?

benlesh replied:
Updated.

mhevery reviewed (changes requested):
I think this PR should also do the changes to the compiler which generates the SVG namespace instructions. Can we add that?

vicb commented on packages/platform-browser/test/dom/dom_renderer_spec.ts:
```diff
@@ -28,23 +28,14 @@ import {NAMESPACE_URIS} from '../../src/dom/dom_renderer';
 
     describe('setAttribute', () => {
       describe('with namespace', () => {
-        it('xmlns', () => shouldSetAttributeWithNs('xmlns'));
```
what is this change ?

benlesh replied:
Previously attributes were being set incorrectly with a `name:likethis`, the `prefix:` is an artifact of HTML serialization of namespaced attributes and not part of the in-memory DOM name of the attribute.  This change just corrects the test.

vicb commented:
I was thinking that we should add more tests to make sure directives are triggered as before (render2) when we have a) NS elements and attributes b) NS selectors.

It's also about making sure this PR is compatible with the recent changes from @pkozlowski-opensource wrt to attributes used for directives only.

mary-poppins commented:
You can preview c8a1615 at https://pr23899-c8a1615.ngbuilds.io/.

mary-poppins commented:
You can preview 0cd257c at https://pr23899-0cd257c.ngbuilds.io/.

JoostK commented on packages/compiler/src/render3/r3_identifiers.ts:
```diff
@@ -39,6 +39,12 @@ export class Identifiers {
   static bind: o.ExternalReference = {name: 'ɵb', moduleName: CORE};
 
+
+  static namespace: o.ExternalReference = {name: 'ɵN', moduleName: CORE};
```
We should add these to `angularCoreEnv` in `packages/core/src/render3/jit/environment.ts` as well.

benlesh replied:
done.

JoostK commented on packages/compiler/src/render3/view/template.ts:
```diff
@@ -23,7 +23,7 @@ import {ParseError, ParseSourceSpan} from '../../parse_util';
 import {CssSelector, SelectorMatcher} from '../../selector';
 import {BindingParser} from '../../template_parser/binding_parser';
-import {OutputContext, error} from '../../util';
+import {OutputContext, error, splitAtColon} from '../../util';
```
No longer needed.

benlesh replied:
done. thanks.

JoostK commented on packages/compiler/src/render3/view/template.ts:
```diff
@@ -381,6 +389,18 @@ export class TemplateDefinitionBuilder implements t.Visitor<void>, LocalResolver
       this.instruction(
           this._creationCode, element.endSourceSpan || element.sourceSpan, R3.elementEnd);
+
+      /* NOTE(benlesh): since we're not keeping a track of what previous namespace we're in,
```
We could easily track a stack of the active namespace, I guess

vicb replied:
agreed. using the call stack should work:
```
visitElement(el) {
  const wasInNamespace = this.namespace;
  const currentNamespace = getNamespace(el);
  this.generateNamespaceChangeIfNeeded(wasInNamespace, currentNameSpace); // should have a better name
  this.namespace = currentNamespace;  // should probably be part of the method above
  // ... recurse
  this.generateNamespaceChangeIfNeeded(currentNameSpace, wasInNamespace);
  this.namespace = wasInNamespace;
}
```

benlesh replied:
I've done something similar now.

JoostK commented on packages/compiler/src/render3/view/template.ts:
```diff
@@ -381,6 +389,18 @@ export class TemplateDefinitionBuilder implements t.Visitor<void>, LocalResolver
+      if (element.name === ':svg:svg' || element.name === ':math:math') {
+        this.instruction(
+            this._creationCode, element.endSourceSpan || element.sourceSpan, R3.namespaceHTML);
+      } else if (element.name === ':svg:foreignObject') {
```
Why is this handled separately from above if? Does the same thing :-)

benlesh replied:
This has now completely changed.

JoostK commented on packages/compiler/test/render3/r3_compiler_compliance_spec.ts:
```diff
@@ -66,6 +66,51 @@ describe('compiler compliance', () => {
+          if (rf & 1) {
+            $r3$.ɵE(0, 'div', $e0_attrs$);
+            $r3$.ɵNS();
+            $r3$.ɵE(1, ':svg:svg');
```
The `:svg:` prefix seems redundant to me.

vicb commented on packages/compiler/src/render3/view/template.ts:
```diff
@@ -320,6 +320,14 @@ export class TemplateDefinitionBuilder implements t.Visitor<void>, LocalResolver
 
     const implicit = o.variable(CONTEXT_NAME);
 
+    if (element.name === ':svg:svg') {
```
What about comparing the namespace with the previous one and generate the instruction when they are different ?

vicb commented on packages/compiler/test/render3/r3_compiler_compliance_spec.ts:
```diff
@@ -66,6 +66,51 @@ describe('compiler compliance', () => {
+            $r3$.ɵE(1, ':svg:svg');
+            $r3$.ɵEe(2, ':svg:circle', $e2_attrs$);
+            $r3$.ɵe();
+            $r3$.ɵNH();
```
Probably we should not generate this one at the end of a template function.
As discussed with @mhevery we need to reset the namespace to HTML before executing a template function

benlesh replied:
Done.

mary-poppins commented:
You can preview c7ddf3b at https://pr23899-c7ddf3b.ngbuilds.io/.

mary-poppins commented:
You can preview 2d21034 at https://pr23899-2d21034.ngbuilds.io/.

vicb commented on packages/core/test/render3/compiler_canonical/template_variables_spec.ts:
```diff
@@ -133,6 +133,57 @@ describe('template variables', () => {
     expect(toHtml(renderComponent(MyComponent))).toEqual('<ul></ul>');
   });
 
+  it('should support a let variable and reference inside of SVG', () => {
```
`it('should support embedded views in the SVG namespace', ...)` ?

benlesh replied:
done

vicb commented on packages/core/test/render3/compiler_canonical/template_variables_spec.ts:
```diff
@@ -133,6 +133,57 @@ describe('template variables', () => {
+    // /NON-NORMATIVE
+
+    // TODO(chuckj): update when the changes to enable ngForOf lands.
+    expect(toHtml(renderComponent(MyComponent))).toEqual('<svg></svg>');
```
looks wrong

benlesh replied:
I agree. This was duplicated from another similar test. I assume, because of the existing comment, that "changed to enable ngForOf" never landed?

vicb commented on packages/core/test/render3/compiler_canonical/template_variables_spec.ts:
```diff
@@ -133,6 +133,57 @@ describe('template variables', () => {
+        [ForOfDirective.ngDirectiveDef];
+    // /NON-NORMATIVE
+
+    // TODO(chuckj): update when the changes to enable ngForOf lands.
```
"chuckj" here in code your code looks wrong !

benlesh replied:
changed to benlesh... I'm still not sure about the test underneath it. Is someone tracking the change that needs to be made? @mhevery

benlesh commented on packages/compiler/src/render3/view/template.ts:
```diff
@@ -328,6 +359,14 @@ export class TemplateDefinitionBuilder implements t.Visitor<void>, LocalResolver
+      // to the HTML namespace inside of it
+      if (element.name === ':svg:foreignObject') {
+        // NOTE(benlesh): this may cause extremem corner-case bugs if someone was to do something
+        //   like <math>...<foreignObject></foreignObject>...</math>.
```
No, what I mean is that if someone puts a `foreignObject` tag inside of a `math` tag, Exiting the `foreignObject` tag will set them into the HTML (null) namespace:

```html
<math>
  <mrow><!-- Valid MathML stuff --></mrow>
  <foreignObject></foreignObject>
  <mrow><!-- this is created as HTMLUnknownElement now --></mrow>
</math>
``` 

I'm not even sure what the behavior would actually be in this case... but it's an _extreme_ corner case. MathML only works in two browsers (Firefox and Safari), and someone would have to put a `foreignObject` tag in it, which is a crazy thing to do.

vicb replied:
It would be easy to restore `MathML` or `SVG` if you compare `this._namespace` with `wasInNamespace` on closing the element as in the snippet I gave there.

benlesh commented on packages/compiler/test/render3/r3_compiler_compliance_spec.ts:
```diff
@@ -1230,6 +1274,78 @@ describe('compiler compliance', () => {
+              if (rf & 2) { $r3$.ɵp(1,'forOf',$r3$.ɵb(ctx.items)); }
+              function MyComponent__svg_g_Template_1(rf:IDENT,ctx0:IDENT) {
+                if (rf & 1) {
+                  $r3$.ɵE(0,':svg:g');
```
Fixed

mary-poppins commented:
You can preview d46b81e at https://pr23899-d46b81e.ngbuilds.io/.

mary-poppins commented:
You can preview 4cb7b1f at https://pr23899-4cb7b1f.ngbuilds.io/.

mhevery commented on packages/compiler/test/render3/r3_compiler_compliance_spec.ts:
```diff
@@ -66,6 +66,50 @@ describe('compiler compliance', () => {
+
+              @Component({
+                selector: 'my-component',
+                template: \`<div class="my-app" title="Hello"><svg><circle cx="50" cy="100" r="25"/></svg></div>\`
```
Add `<p>` after `<svg>` to prove that we go back to `HTML` namespace.

benlesh replied:
This is done in another test now.

mary-poppins commented:
You can preview 429edb0 at https://pr23899-429edb0.ngbuilds.io/.

mary-poppins commented:
You can preview 4b4dfbe at https://pr23899-4b4dfbe.ngbuilds.io/.

mary-poppins commented:
You can preview 5d1a8f9 at https://pr23899-5d1a8f9.ngbuilds.io/.

vicb commented on packages/core/src/render3/instructions.ts:
```diff
@@ -586,7 +629,11 @@ export function elementStart(
-  const native: RElement = renderer.createElement(name);
+
+  const native: RElement = _currentNamespace === null || isProceduralRenderer(renderer) ?
+      renderer.createElement(name) :
```
Shouldn't you pass the namespace as the 2nd arg here ?

benlesh replied:
No.

If it's a procedural renderer, it doesn't have a namespace exposed on the API... and if there's no namespace, we're supposed to use `document.createElement` which doesn't have a namespace argument.

benlesh replied:
Okay, fixed per our discussion.

mary-poppins commented:
You can preview 052fb5e at https://pr23899-052fb5e.ngbuilds.io/.

vicb reviewed (approved):
Looks great !

mary-poppins commented:
You can preview 9e0a117 at https://pr23899-9e0a117.ngbuilds.io/.

vicb commented:
blocked: rebase needed

mary-poppins commented:
You can preview c475618 at https://pr23899-c475618.ngbuilds.io/.

vicb commented:
force merging: g3 was green before rebase

mary-poppins commented:
You can preview 3b1b932 at https://pr23899-3b1b932.ngbuilds.io/.

vicb commented:
causes a breakage in g3

mary-poppins commented:
You can preview 4e8c11f at https://pr23899-4e8c11f.ngbuilds.io/.

vicb commented on packages/compiler/src/render3/view/template.ts:
```diff
@@ -265,6 +271,11 @@ export class TemplateDefinitionBuilder implements t.Visitor<void>, LocalResolver
       const value = attr.value;
+
+      if (name.startsWith(':xmlns:')) {
+        continue;
```
You should rather add a branch to the if below ?

vicb replied:
This comment has not been addressed.
However now that I understand better what you are trying to do, this change is probably not the most appropriate.

You should probably rather update the [r3_template_transformer](https://github.com/angular/angular/blob/a6e797b8f53132c3b249ff16f1d3c2bed759b08c/packages/compiler/src/render3/r3_template_transform.ts#L179) and filter out attributes with the `xmlns` NS. That piece of code transforms the HTML AST into an R3 AST. As R3 does nothing with those attributes, they should not be part of the AST.

vicb reviewed (dismissed):
- Please use fixup commits when modifying an existing PRs - it makes it much easier to see the updates,
- I am not sure a test has been added ? 
- I am not sure what got fixed for render 2 ?
- please run a global presubmit

mary-poppins commented:
You can preview 6861181 at https://pr23899-6861181.ngbuilds.io/.

benlesh commented:
@vicb ... Github ate my review dismissal message.

- I squashed the commits at @mhevery's request.
- I just added a piece that ignores attributes prefixed with `:xmlns:` while visiting the element.
- Test exists [here](https://github.com/angular/angular/pull/23899/files#diff-b9d784499e486a026b7e92e7a1ee99eaR69)
- Presubmit in google3 was run.

mary-poppins commented:
You can preview db27a4c at https://pr23899-db27a4c.ngbuilds.io/.

vicb commented on packages/compiler/test/render3/r3_compiler_compliance_spec.ts:
```diff
@@ -66,6 +66,137 @@ describe('compiler compliance', () => {
       expectEmit(result.source, template, 'Incorrect template');
     });
 
+    it('should skip adding instructions for xmlns attributes', () => {
```
- Could you reword this ? there is no "instruction" for attributes,
- The g3 failure is 1) a runtime failure with 2) the render2 - this is testing ivy compiler which is a different code path.

-> I think you should add some integration test for the render 2

vicb reviewed (changes requested):
- your fix is for ivy while the bug is with render2,
- the ivy fix is not optimal

-> add a test for render 2
-> improve the fix for ivy
-> run a GLOBAL presubmit (this one should have failed with the current code) and add a link in the PR.
Thanks

mhevery commented:
superseded by: https://github.com/angular/angular/pull/24386, https://github.com/angular/angular/pull/24379, https://github.com/angular/angular/pull/24377.
